### PR TITLE
test: add tests for associated type multiple `Eff`s and types

### PR DESCRIPTION
**Pull request by jaschdoc**
- [x] Add traits
- [x] Add instance for Int32.
- [x] Add instance for String
- [x] Add instance for Vector[a]
- [ ] Write polymorphic functions that use the trait.
- [ ] Write monomorphic tests that exercise the trait directly and also the polymorphic functions.
- [ ] Add tests

Related to https://github.com/flix/flix/issues/7424

### CODE REVIEW

**Comment by jaschdoc** [comment]
I'm not sure this should be a type error?

```scala
mod Test.Dec.Assoc.Type.Eff.Multiple {

    eff Observe {
        pub def capture(x: Int32): Unit
    }

    enum Maybe[a] {
        case Nothing,
        case Just(a)
    }

    enum Split[e: Eff, a: Type] {
        case Nothing,
        case Value(a),
        case Pending(a -> Unit \ e, a)
    }

    trait Composable[a] {
        pub type E: Eff
        pub type B: Type -> Type
        pub type C: Eff -> Type -> Type
        pub def compose(f: a -> Unit \ Composable.E[a], x: a): Composable.C[a][Composable.E[a], Composable.B[a][a]]
    }

    instance Composable[Int32] {
        pub type E = Observe
        pub type B = Maybe
        pub type C = Split

        pub def compose(f: Int32 -> Unit \ Observe, x: Int32): Split[Observe, Maybe[Int32]] =
            Split.Pending(f, Maybe.Just(x))

    }
}
```
```
❌ -- Type Error -------------------------------------------------- file://../TestEffMultiple.flix

>> Unable to unify the types: 'Int32' and 'Maybe[Int32]'.

44 |             Split.Pending(f, Maybe.Just(x))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                 mismatched types.

Type One: ((Int32 -> Unit \ Observe, Int32)) -> Split[Observe, Int32]
Type Two: ((Int32 -> Unit \ Observe, Maybe[Int32])) -> Split[Observe, Maybe[Int32]]
```

**Comment by jaschdoc** [comment]
Should this be a type error?

```scala
mod Test.Dec.Assoc.Type.Eff.Multiple {

    ////////////////////////////////////////////////////////
    // Type Definitions                                   //
    ////////////////////////////////////////////////////////

    eff Observe {
        pub def captureInt32(x: Int32): Unit
        pub def captureString(x: String): Unit
    }

    enum Split[e: Eff, a: Type] {
        case Nothing,
        case Value(a),
        case Pending(a -> Unit \ e, a)
    }

    trait Composable[a] {
        pub type E: Eff
        pub type B: Type -> Type
        pub type C: Eff -> Type -> Type
        pub def compose(f: a -> Unit \ Composable.E[a], x: a): Composable.C[a][Composable.E[a], Composable.B[a][a]]
        pub def reduce(x: Composable.C[a][Composable.E[a], Composable.B[a][a]]): Composable.B[a][a] \ Composable.E[a]
    }

    instance Composable[Vector[a]] with Composable[a] {
        pub type E = Composable.E[a]
        pub type B = Vector
        pub type C = Split

        pub def compose(f: Vector[a] -> Unit \ Composable.E[a], x: Vector[a]): Split[Composable.E[a], Vector[a]] =
            Split.Pending(f, x)

        pub def reduce(x: Split[Composable.E[a], Vector[a]]): Vector[a] \ Composable.E[a] = match x {
            case Split.Nothing       => Vector#{}
            case Split.Value(v)      => v
            case Split.Pending(f, v) => f(v); v
        }
    }
}
```
```
❌ -- Instance Error -------------------------------------------------- file://.../TestEffMultiple.flix


Mismatched signature 'compose' required by 'Composable'.

103 |         pub def compose(f: Vector[a] -> Unit \ Composable.E[a], x: Vector[a]): Split[Composable.E[a], Vector[a]] =
                      ^^^^^^^
                      mismatched signature.

Expected scheme: ∀(a). (Vector[a] -> Unit \ E[Vector[a]]) -> (Vector[a] -> C[Vector[a], E[Vector[a]], B[Vector[a], Vector[a]]]) with Test.Dec.Assoc.Type.Eff.Multiple.Composable[Vector[a]]
Actual scheme:   (Vector[a] -> Unit \ E[a]) -> (Vector[a] -> Split[E[a], Vector[a]]) with Test.Dec.Assoc.Type.Eff.Multiple.Composable[Vector[a]]

Explanation:
Tip: Modify the definition to match the signature.
❌ -- Instance Error -------------------------------------------------- file://.../TestEffMultiple.flix


Mismatched signature 'reduce' required by 'Composable'.

106 |         pub def reduce(x: Split[Composable.E[a], Vector[a]]): Vector[a] \ Composable.E[a] = match x {
                      ^^^^^^
                      mismatched signature.

Expected scheme: ∀(a). C[Vector[a], E[Vector[a]], B[Vector[a], Vector[a]]] -> B[Vector[a], Vector[a]] \ E[Vector[a]] with Test.Dec.Assoc.Type.Eff.Multiple.Composable[Vector[a]]
Actual scheme:   Split[E[a], Vector[a]] -> Vector[a] \ E[a] with Test.Dec.Assoc.Type.Eff.Multiple.Composable[Vector[a]]

Explanation:
Tip: Modify the definition to match the signature.
```

**Comment by mlutze** [comment]
I think it is because of the `Composable.B[a][a]`. This substitutes to `Composable.B[Vector[a], Vector[a]]`, which reduces to `Vector[Vector[a]]` but you just have `Vector[a]`.

**Comment by jaschdoc** [comment]
Ah thanks! :)

**Comment by jaschdoc** [comment]
Okay so now it says that it is an irreducible associated type. Is it because of the recurrence of `Vector` when defining the instance for `Vector`? Not sure if this is also an error, but every instance in my program is now an irreducible associated type?

**Comment by mlutze** [comment]
> Okay so now it says that it is an irreducible associated type. Is it because of the recurrence of `Vector` when defining the instance for `Vector`? Not sure if this is also an error, but every instance in my program is now an irreducible associated type?

Irreducible shouldn't happen anymore, so it's probably a bug. Can you post the new program and the error message?

**Comment by jaschdoc** [comment]
Ah I was on latest release rather than `master` :/

File: main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    }
+
+
+    ////////////////////////////////////////////////////////
+    // Polymorphic functions using Transformable          //
+    ////////////////////////////////////////////////////////
+
+    pub def transformAndCheckEq(eq: Transformable.B[a], x: a): Bool \ Transformable.E[a] with Transformable[a], Eq[Transformable.B[a]] =
+        Transformable.transform(x) == eq
```
Comment (jaschdoc) [inline main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:119]: Is this `Eq` bound allowed? @mlutze

Comment (mlutze) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:119]: No (I think), but you can hack it with an equality constraint: `... Eq[b] where Assoc[a] ~ b`

**Comment by magnus-madsen** [comment]
@jaschdoc Is this blocked or OK?

**Comment by jaschdoc** [comment]
Error
```scala
    trait Transformable[a] {
        pub type E: Eff
        pub type B: Type
        pub def transform(x: a): Transformable.B[a] \ Transformable.E[a]
    }

    pub def transformAndCheckEq[b](eq: Transformable.B[a], x: a): Bool \ Transformable.E[a] with  Transformable[a], Eq[b] where Transformable.B[a] ~ b =
        Transformable.transform(x) == eq

```

```
#
# An unexpected error has been detected by the Flix compiler:
#
#   illegal type: B[Error]
#
# This is a bug in the Flix compiler. Please report it here:
#
# https://github.com/flix/flix/issues
#
# -- Flix Compiler --
#
# Flix Version : 0.46.0
#   incremental: All
#
# -- Java Virtual Machine --
#
# JVM Version  : 21.0.1 (2023-10-17)
# JVM Vendor   : GraalVM Community
# JAVA_HOME    : /home/_/.sdkman/candidates/java/21.0.1-graalce
# System       : Linux
#
# -- Stack Trace --
ca.uwaterloo.flix.util.InternalCompilerException: illegal type: B[Error]
	at ca.uwaterloo.flix.language.phase.Kinder$.$anonfun$checkAssocTypes$5(Kinder.scala:385)
	at ca.uwaterloo.flix.util.Validation$.$anonfun$fastTraverse$1(Validation.scala:284)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at ca.uwaterloo.flix.util.Validation$.fastTraverse(Validation.scala:283)
	at ca.uwaterloo.flix.util.Validation$.traverse(Validation.scala:237)
	at ca.uwaterloo.flix.util.Validation$.traverseX(Validation.scala:264)
	at ca.uwaterloo.flix.language.phase.Kinder$.checkAssocTypes(Kinder.scala:371)
	at ca.uwaterloo.flix.language.phase.Kinder$.$anonfun$visitSpec$5(Kinder.scala:343)
	at ca.uwaterloo.flix.util.Validation$.$anonfun$curry$10(Validation.scala:388)
	at ca.uwaterloo.flix.util.Validation$.ap(Validation.scala:337)
	at ca.uwaterloo.flix.util.Validation$.flatMapN(Validation.scala:569)
	at ca.uwaterloo.flix.language.phase.Kinder$.visitSpec(Kinder.scala:337)
	at ca.uwaterloo.flix.language.phase.Kinder$.$anonfun$visitDef$1(Kinder.scala:283)
	at ca.uwaterloo.flix.util.Validation$.flatMapN(Validation.scala:524)
	at ca.uwaterloo.flix.language.phase.Kinder$.visitDef(Kinder.scala:281)
	at ca.uwaterloo.flix.language.phase.Kinder$.$anonfun$visitDefs$1(Kinder.scala:268)
	at ca.uwaterloo.flix.util.ParOps$.$anonfun$parTraverseValues$1(ParOps.scala:106)
	at ca.uwaterloo.flix.util.ParOps$.$anonfun$parMap$2(ParOps.scala:64)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

**Comment by jaschdoc** [comment]
Error
```scala
    trait Composable[a] {
        pub type E: Eff
        pub type B: Type -> Type
        pub type C: Eff -> Type -> Type
        pub def compose(f: a -> Unit \ Composable.E[a], x: a): Composable.C[a][Composable.E[a], Composable.B[a][a]]
        pub def reduce(x: Composable.C[a][Composable.E[a], Composable.B[a][a]]): Composable.B[a][a] \ Composable.E[a]
    }
    pub def composeAndReduceEq[a: Type](eq: Composable.B[a][a], f: a -> Unit \ Composable.E[a], x: a): Bool \ Composable.E[a] with Composable[a], Eq[a] where Composable.B[a] ~ a =
        Composable.reduce(Composable.compose(f, x)) == eq

```
```
#
# An unexpected error has been detected by the Flix compiler:
#
#   Illegal kind: 'Type' of type 't879401'.
#
# This is a bug in the Flix compiler. Please report it here:
#
# https://github.com/flix/flix/issues
#
# -- Flix Compiler --
#
# Flix Version : 0.46.0
#   incremental: All
#
# -- Java Virtual Machine --
#
# JVM Version  : 21.0.1 (2023-10-17)
# JVM Vendor   : GraalVM Community
# JAVA_HOME    : /home/_/.sdkman/candidates/java/21.0.1-graalce
# System       : Linux
#
# -- Stack Trace --
ca.uwaterloo.flix.util.InternalCompilerException: Illegal kind: 'Type' of type 't879401'.
	at ca.uwaterloo.flix.language.ast.Type$Apply.kind$lzycompute(Type.scala:494)
	at ca.uwaterloo.flix.language.ast.Type$Apply.kind(Type.scala:491)
	at ca.uwaterloo.flix.language.phase.unification.Unification$.unifyTypes(Unification.scala:87)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.$anonfun$resolveTraitConstraint$4(ConstraintSolver.scala:468)
	at scala.collection.Iterator$$anon$9.next(Iterator.scala:584)
	at scala.collection.IterableOnceOps.collectFirst(IterableOnce.scala:1142)
	at scala.collection.IterableOnceOps.collectFirst$(IterableOnce.scala:1134)
	at scala.collection.AbstractIterator.collectFirst(Iterator.scala:1300)
	at ca.uwaterloo.flix.util.collection.ListOps$.findMap(ListOps.scala:38)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.$anonfun$resolveTraitConstraint$1(ConstraintSolver.scala:467)
	at ca.uwaterloo.flix.util.Result.flatMap(Result.scala:62)
	at ca.uwaterloo.flix.util.Result.flatMap$(Result.scala:61)
	at ca.uwaterloo.flix.util.Result$Ok.flatMap(Result.scala:94)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolveTraitConstraint(ConstraintSolver.scala:450)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolveOne(ConstraintSolver.scala:275)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.tryResolve$1(ConstraintSolver.scala:250)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolveOneOf(ConstraintSolver.scala:260)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolve(ConstraintSolver.scala:226)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.visitSpec(ConstraintSolver.scala:110)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.visitDef(ConstraintSolver.scala:56)
	at ca.uwaterloo.flix.language.phase.Typer$.visitDef(Typer.scala:177)
	at ca.uwaterloo.flix.language.phase.Typer$.$anonfun$visitDefs$2(Typer.scala:162)
	at ca.uwaterloo.flix.util.ParOps$.$anonfun$parTraverseValues$1(ParOps.scala:106)
	at ca.uwaterloo.flix.util.ParOps$.$anonfun$parMap$2(ParOps.scala:64)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

File: main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    pub def transformAndCheckEq(eq: Transformable.B[a], x: a): Bool \ Transformable.E[a] with Transformable[a], Eq[a] where Transformable.B[a] ~ a =
+        Transformable.transform(x) == eq
+
+
+    ////////////////////////////////////////////////////////
+    // Polymorphic functions using Composable             //
+    ////////////////////////////////////////////////////////
+
+    pub def composeAndReduceEq(eq: Composable.B[a][a], f: a -> Unit \ Composable.E[a], x: a): Bool \ Composable.E[a] with Composable[a], Eq[a] where Composable.B[a][a] ~ a =
```
[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: So is it impossible to do what I'm trying to achieve?

[the remark this answers is not in the document]
Comment (mlutze) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: At least for that constraint, yes.

Does this not work without the constraint?

[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: I need equality on the type `C` where `C` is the associated type of an instance for `B` which again is an associated type for an instance for `A` if that makes sense?

[the remark this answers is not in the document]
Comment (mlutze) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: > I need equality on the type `C` where `C` is the associated type of an instance for `B` which again is an associated type for an instance for `A` if that makes sense?

It does not :D

[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: Ok well I'm trying to use the trait twice so I need `Eq` on the second return type of the trait. Maybe the code makes sense?

[the remark this answers is not in the document]
Comment (mlutze) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: I'm not sure what "second return type" means. But if you want equality on `Composable.B[a][a]` then I think you can do `with Eq[blah] where Composable.B[a][a] ~ blah` where `blah` is some type variable not used anywhere else.

[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: Yeah but that gives unused type variable error 😀

[the remark this answers is not in the document]
Comment (mlutze) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: oh noooooo that's a bug

[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: That makes me happy haha

[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: Ah wait

[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]:
```
❌ -- Syntax Error -------------------------------------------------- file:///.../TestEffMultiple.flix

>> Illegal equality constraint.

126 |     pub def composeAndReduceEq(eq: Composable.B[a][a], f: a -> Unit \ Composable.E[a], x: a): Bool \ Composable.E[a] with Composable[a], Eq[b] where Composable.B[a][a] ~ b =
                                                                                                                                                           ^^^^^^^^^^^^^^^^^^^^^^
                                                                                                                                                           Equality constraints must have the form: `Assoc[var] ~ Type`.
```

[the remark this answers is not in the document]
Comment (mlutze) [reply main/test/flix/Test.Dec.Assoc.Type.Eff.Multiple.flix:126]: Ah, that's a trickier problem. We can't currently do a constraint `Eq[X[a]]` if X is abstract. It's a design question. See https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Functor-Classes.html#t:Eq1

**Comment by magnus-madsen** [comment]
@jaschdoc @mlutze What's this PR about and what should happen to it?

**Comment by jaschdoc** [comment]
I don't know how urgent it is for Matt, but this is just lots of tests for associated types and effects that he asked me to add an some point. Then we found some broken things here and there but I don't know if they have been fixed since. I can try to look into it again

**Comment by magnus-madsen** [comment]
Possibly we want to wait for set effects to arrive with the new typer.

**Comment by magnus-madsen** [comment]
I am closing this. If you want, you can revisit it.

**Pull request closed without merging**